algebird-core/src/main/scala/com/twitter/algebird/experimental/BloomFilter.scala

```diff
@@ -0,0 +1,413 @@
+
+    /**
+     * The number of bits set to true
+     */
+    override def numBits: Int = bits.size.toInt
+
+    override val toBitSet: BitSet = bits
+
+    override def ++(other: Hash): Hash =
+      other match {
+        case Empty               => this
+        case Item(item)          => Instance(bits | BitSet(hashToArray(item)))
+        case Instance(otherBits) => Instance(bits | otherBits)
+      }
+
+    override def +(item: A): Hash = this ++ Item(item)
+
+    override def checkAndAdd(other: A): (Hash, ApproximateBoolean) =
+      (this + other, contains(other))
+
+    override def maybeContains(item: A): Boolean = {
+      val il = hashToArray(item)
+      var idx = 0
+      var found = true
+      while (idx < il.length && found) {
+        val i = il(idx)
+        if (!bits(i)) {
+          found = false
+        }
+        idx += 1
+      }
+      found
+    }
+
+    // use an approximation width of 0.05
+    override def size: Approximate[Long] =
+      BloomFilter.sizeEstimate(numBits, numHashes, width, 0.05)
+  }
+
+  implicit val monoid: Monoid[Hash] with BoundedSemilattice[Hash] =
+    new Monoid[Hash] with BoundedSemilattice[Hash] {
+      override val zero: Hash = Empty
+
+      /**
+       * Assume the bloom filters are compatible (same width and same hashing functions).  This
+       * is the union of the 2 bloom filters.
+       */
+      override def plus(left: Hash, right: Hash): Hash = left ++ right
+
+      override def sum(t: TraversableOnce[Hash]): Hash =
+        if (t.iterator.isEmpty) empty
+        else {
+          var bfItem: Item = null
+          val iter = t.iterator
+          var counter = 0
+          var bs = BitSet.newEmpty(0)
+
+          while (iter.hasNext) {
+            iter.next() match {
+              case Empty => ()
+              case bi @ Item(item) =>
+                bfItem = bi
+                val array = hashToArray(item)
+                counter += array.length
+                bs = bs.mutableAdd(array)
+              case Instance(bitset) =>
+                val iter = bitset.iterator
+                while (iter.hasNext) {
+                  counter += 1
+                  bs = bs.mutableAdd(iter.next())
+                }
+            }
+          }
+
+          if (counter == 0) empty
+          else if (counter == n && (bfItem != null)) bfItem
+          else Instance(bs)
+        }
+
+      override def sumOption(t: TraversableOnce[Hash]): Option[Hash] =
+        if (t.iterator.isEmpty) None else Some(sum(t))
+    }
+
+  val aggregator: MonoidAggregator[A, Hash, Hash] = new MonoidAggregator[A, Hash, Hash] {
+    override val monoid: Monoid[Hash] = self.monoid
+
+    override def prepare(value: A): Hash = Item(value)
+
+    override def present(bf: Hash): Hash = bf
+  }
+
+  implicit val equiv: Equiv[Hash] = new Equiv[Hash] {
+    override def equiv(a: Hash, b: Hash): Boolean =
+      (a eq b) || ((a.numHashes == b.numHashes) &&
```

Comment:
we don't need to check numHashes or width right? the path dependent type verifies that statically, no?

Reply:
right! missed this